HDD and SSD used bytes counters

jepett0 opened the pull request:
#3275

In this PR two new counters are added:
- `resources.storage.used_bytes.ssd`
- `resources.storage.used_bytes.hdd`

Necessary subcounters like `resources.storage.used_bytes.table.ssd` (note the added `table` part) are also added. These counters are needed to monitor SSD and HDD storage usage separately. 

In the follow-up [PR](https://github.com/ydb-platform/ydb/pull/4765) `.ssd` and `.hdd` versions of `resources.storage.limit_bytes` are added too.

Probably the most interesting part of this PR is the "hacky" [mapping](https://github.com/ydb-platform/ydb/blob/4fca557e2646a1de873613694c2adfd10c780b2f/ydb/core/tx/schemeshard/schemeshard_info_types.cpp#L54) of storage pool kinds to `ssd` and `hdd` counters. It is discussed in more detail in these comments:
- https://github.com/ydb-platform/ydb/pull/4727#discussion_r1608650304
- https://github.com/ydb-platform/ydb/pull/4727#discussion_r1608656793

The data for the `used_bytes` counters is sourced from the SchemeShard's counters:
- `SchemeShard/DiskSpaceTablesTotalBytesOnSsd`
- `SchemeShard/DiskSpaceTablesTotalBytesOnHdd`

which are updated each time the SchemeShard receives TEvPeriodicTableStats from DataShards.

Currently, only row-oriented tables report disk space usage separately by channels, so effectively:
- `resources.storage.used_bytes.ssd` is equal to `resources.storage.table.used_bytes.ssd` (note the added `table` part)
- `resources.storage.used_bytes.hdd` is equal to `resources.storage.table.used_bytes.hdd` (note the added `table` part)

In the future, we will start accounting for HDD and SSD usage separately for both column-oriented tables and topics. Consequently, `resources.storage.used_bytes.ssd` will more accurately represent the total SSD space used by all entities in the database.

P.S. For the best reviewing experience, I recommend reviewing the [commits](https://github.com/ydb-platform/ydb/pull/4727/commits) individually.

jepett0 committed:
HDD and SSD storage counters

jepett0 committed:
Minor refactoring + log level increase in some unit tests

Log level needs to be TRACE to see the raw table stats received by the SchemeShard

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_info_types.cpp:
```diff
@@ -1577,19 +1577,21 @@ void TAggregatedStats::UpdateShardStats(TShardIdx datashardIdx, const TPartition
             aggregatedStoragePoolStats->DataSize += newStoragePoolStats.DataSize - (oldStoragePoolStats ? oldStoragePoolStats->DataSize : 0u);
             aggregatedStoragePoolStats->IndexSize += newStoragePoolStats.IndexSize - (oldStoragePoolStats ? oldStoragePoolStats->IndexSize : 0u);
         } else {
+            Y_ABORT_UNLESS(!oldStoragePoolStats, "Old stats are present, but they haven't been aggregated.");
```
I added this Y_ABORT_UNLESS for greater similarity with the for loop below (which I have added before).

I believe that the state of the TAggregatedStats in which:
1) the aggregated stats do not contain an entry for a particular storage pool kind
2) there is an old stats entry for this storage kind

cannot be correctly handled silently. The semantics of TEvPeriodicStats messages implies that the aggregated stats are incremented by `newStats - oldStats (or 0 if they are absent)`. If some old stats, that haven't been aggregated, are present, it signals an error in aggregation, which should be investigated.

I suppose it is better to discover these kinds of errors during testing deployment than to silently ignore and calculate an aggregate using some stale oldStats.

ijon replied:
But also we don't want to lose entire node (with schemeshard and all other tablets) to a cycle of the same aborts. `Y_ABORT_UNLESS` are too risky in that regard.

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_info_types.cpp:
```diff
@@ -50,6 +51,16 @@ EDiskUsageStatus CheckStoragePoolsQuotas(const THashMap<TString, TStoragePoolUsa
 }
 
+EUserFacingStorageType GetUserFacingStorageType(const TString& poolKind) {
+    if (poolKind.StartsWith("ssd")) {
```
This should cover "ssd" and "ssdencrypted". This hack will certainly be documented.

ijon replied:
Also there should be LOUD (but short if possible) indication and explanation that this is a workaround.
That makes sense only in a limited case when there are no multiple storage-pools with the same kind and storage-pool kinds are all in (ssh*, hdd*, rot*, nvme*) set.

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_info_types.h:
```diff
@@ -1281,6 +1281,12 @@ struct TSchemeQuotas : public TVector<TSchemeQuota> {
     mutable size_t LastKnownSize = 0;
 };
 
+enum class EUserFacingStorageType {
```
This enum highlights that we categorize numerous storage pool kinds into just two reported storage types: HDD and SSD.

jepett0 committed:
Map nvme -> ssd, rot -> hdd for used_bytes counters

ijon commented on ydb/core/tx/schemeshard/schemeshard__operation_common_subdomain.h:
```diff
@@ -6,33 +6,43 @@
-                                    const TVector<TStoragePool>& pools,
-                                    const TString& path,
-                                    TString& error
+inline bool CheckStoragePoolsInQuotas(const Ydb::Cms::DatabaseQuotas& quotas,
+                                      const TVector<TStoragePool>& pools,
+                                      const TString& path,
+                                      TString& error
```
No need to align args to the opening brace, double indentation from the left is enough.

jepett0 replied:
I hope I got your idea right

jepett0 committed:
Review fixes

- fix style
- add descriptive comment about the nature of the storage pool kind mapping

jepett0 committed:
Review fix: remove ABORT_UNLESS for a minor problem
